### ISSUE TITLE
Make QueryPartInternal.toSQL() and bind() final in favour of the new accept() method

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#3323]
A variety of improvements were implemented in jOOQ 3.4 (#2694) to simplify `QueryPart` traversal. Instead of separating SQL rendering and variable binding methods, a single `accept()` method was introduced, which can handle both.

The deprecated `toSQL(RenderContext)` and `bind(BindContext)` methods will be removed in jOOQ 4.0. `CustomQueryPart` implementations should adapt until then.

**Comment by ams-tschoening** [comment]
I'm in progress of updating to jOOQ 3.11 from jOOQ 3.6 and having trouble with this change. I have a custom field which formerly implemented `toSQL` and `bind` and everything worked, now after moving to `accept` that method always gets called with `DefaultRenderContext` only. It never gets called with some `BindContext` and therefore no values for indexed parameters reach the JDBC driver and it throws exceptions therefore.

So, what is the expected behaviour in my case? Should `accept` have been called with both contexts one after another by default? Or have some other things changed around custom fields and I need to do additional things to get some `BindContext`?

All of your docs only tell to migrate the former implementation of `toSQL` and `bind` to `accept` and that approach has been [deleted from the sources](https://github.com/jOOQ/jOOQ/commit/e6f9368a39d8e950e6832f074d93cff4f73d469f#diff-24c660abc4ce0a3db343200a07179c8fL105) only, but my class isn't called with `BindContext` anymore.

**Comment by lukaseder** [comment]
Thank you very much for your message, @ams-tschoening 

Historically, the distinction between `toSQL()` and `bind()` meant that the jOOQ `QueryPart` expression tree needed to be traversed twice, causing significant performance overhead in benchmarks. The move to a single `accept()` method improved this situation, and made maintaining `QueryPart` implementations much easier. Indeed, there is still a `RenderContext` instance being passed around, but it now also collects the bind variables in a single pass.

I just realised the manual is outdated. I will fix this separately: #8048

For the time being, here's a simple unit test we have that shows how you could implement `accept()`:

```java
    @Test
    public void testCustomQueryPart() throws Exception {
        QueryPart p = new CustomQueryPart() {

            @Override
            public void accept(Context<?> ctx) {
                ctx.visit(val("abc"));
            }
        };

        assertEquals("(x = ?)", r_ref().render(DSL.condition("x = {0}", p)));
        assertEquals("(x = :1)", r_refP().render(DSL.condition("x = {0}", p)));
        assertEquals("(x = 'abc')", r_refI().render(DSL.condition("x = {0}", p)));

        context.checking(new Expectations() {{
            oneOf(statement).setString(1, "abc");
        }});

        int i = b_ref().visit(DSL.condition("x = {0}", p)).peekIndex();
        assertEquals(2, i);

        context.assertIsSatisfied();
    }
```

In other words, as before, `CustomQueryPart` instances are allowed to produce further expression trees, delegating rendering and variable binding to them. In your case, it wouldn't be a bad idea to simply call `ctx.visit(Param)` and delegate both the rendering of the `?` marker as well as the binding of the variable to the `Param`.

As you can see, this change has really simplified jOOQ's internals.

**Comment by ams-tschoening** [comment]
Thanks, I had the feeling that calling `visit` myself is wrong, but it simply works as you said.

**Comment by lukaseder** [comment]
Interesting, what made you think calling `visit` was wrong? (You shouldn't call `accept()` indeed, but `visit()` is fine)

**Comment by ams-tschoening** [comment]
It was just a feeling: I used `bind` and `toSQL` only before with different contexts, now you talked of using `accept` and things like `visit` simply didn't fit from my understanding. I focussed on `bindValue` of `RenderContext` instead, because I was missing a replacement for my former `bind`-method.

I don't think it's worth thinking about this any longer, your documentation seems pretty clear to me now. Thanks!

**Comment by lukaseder** [comment]
OK, perfect. Thanks a lot for your feedback!
